src/sources/docker.rs

```diff
@@ -0,0 +1,888 @@
+                            tokio::spawn(
+                                main.send(info)
+                                    .map_err(|e| error!(message="Unable to return ContainerLogInfo to main",%e))
+                                    .map(|_| ()),
+                            );
+                            Ok(Async::Ready(None))
+                        }
+                        Ok(Async::NotReady) => Ok(Async::NotReady),
+                        Err(error) => {
+                            error!(message = "docker API container logging error",%error);
+                            Err(())
+                        }
+                    };
+                }
+            }
+
+            Ok(Async::Ready(None))
+        })
+        .forward(self.out.clone().sink_map_err(|_| ()))
+        .map(|_| ());
+
+        // Run event_stream
+        tokio::spawn(event_stream);
+    }
+}
+
+/// Kept by main to keep track of container state
+struct ContainerState {
+    /// None if there is a event_stream of this container.
+    info: Option<ContainerLogInfo>,
+    /// True if Container is currently running
+    running: bool,
+    /// Of running
+    generation: u64,
+}
+
+impl ContainerState {
+    /// Container docker ID
+    /// Unix timestamp of event which created this struct
+    fn new(id: String, created: i64) -> Self {
+        let info = ContainerLogInfo {
+            id: id,
+            created,
+            last_log: None,
+            generation: 0,
+        };
+        ContainerState {
+            info: Some(info),
+            running: true,
+            generation: 0,
+        }
+    }
+
+    fn running(&mut self) {
+        self.running = true;
+        self.generation += 1;
+    }
+
+    fn stoped(&mut self) {
+        self.running = false;
+    }
+
+    /// True if it needs to be restarted.
+    #[must_use]
+    fn return_info(&mut self, info: ContainerLogInfo) -> bool {
+        debug_assert!(self.info.is_none());
+        // Generation is the only one strictly necessary,
+        // but with v.running, restarting event_stream is automtically done.
+        let restart = self.running || info.generation < self.generation;
+        self.info = Some(info);
+        restart
+    }
+
+    fn take_info(&mut self) -> Option<ContainerLogInfo> {
+        self.info.take().map(|mut info| {
+            // Update info
+            info.generation = self.generation;
+            info
+        })
+    }
+}
+
+/// Exchanged between main future and event_stream futures
+struct ContainerLogInfo {
+    /// Container docker ID
+    id: String,
+    /// Unix timestamp of event which created this struct
+    created: i64,
+    /// Timestamp of last log message with it's generation
+    last_log: Option<(DateTime<FixedOffset>, u64)>,
+    /// generation of ContainerState at event_stream creation
+    generation: u64,
+}
+
+impl ContainerLogInfo {
+    /// Only logs after or equal to this point need to be fetched
+    #[allow(dead_code)]
```

Comment:
why do we need dead_code here?

Reply:
It's a stray from times when filtering by timestamp wasn't used. It isn't necessary anymore.

Reply:
Scratch that. It's still necessary since filtering by timestamp still isn't used.

Reply:
Is this something we will add later, if so can we add that in a separate PR?

Reply:
Yes, we can add it once `shiplift` releases it.